jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentSourceRange.java

```diff
@@ -0,0 +1,187 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io.internal;
+
+import java.nio.channels.ClosedChannelException;
+
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.util.TypeUtil;
+
+/**
+ * A {@link Content.Source} that provides a range of content from another {@link Content.Source}.
+ */
+public class ContentSourceRange implements Content.Source
```

Comment:
I like that this wrapper allows to have the bounds checks done in a single place. But I do not like that it introduces a concern represented by the `_readToEof` flag controlling if the wrapped resource should be depleted or not. I fear it would be too easy to introduce a resource leak by using this wrapper and not considering the flag whose default to `false` would create a leak if the wrapped resource is not lifecycled separately.

Reply:
If this class doesn't read to EOF it fails the source.  If that leaks, then that is a fault of the source and not this class.